### PR TITLE
Implement locale message caching.

### LINKED ISSUE #19: Implement message dictionary caching.
When `dojo-i18n` is incorporated into [`dojo-widgets`](https://github.com/dojo/widgets), there will need to be a way to synchronously load message dictionaries, in order to reduce the number of `invalidate` calls needed to render the correct locale-specific messages. Since the [`dojo-loader`](https://github.com/dojo/loader) caches modules it has already loaded, creating a separate cache within `dojo-i18n` seemed redundant. However, now that there is a need for synchronous message loading, implementing a cache is also now necessary.

### PR DESCRIPTION
**Pull request by mwistrand**
Resolves #19. As noted in the issue, the purpose for this PR is to facilitate incorporation into `dojo-widgets`.

### CODE REVIEW

File: src/i18n.ts [hunk 1]
```diff
@@ -267,6 +268,29 @@ function validatePath(path: string): void {
...
+ * @return The cached messages object, if it exists.
+ */
+export function getCachedMessages<T extends Messages>(bundlePath: string, locale: string): T | void;
+export function getCachedMessages<T extends Messages>(bundle: Bundle<T>, locale: string): T | void;
+export function getCachedMessages<T extends Messages>(bundle: any, locale: string): T | void {
+	const bundlePath = (typeof bundle === 'string') ? bundle : bundle.bundlePath;
+	const cached = bundleMap[bundlePath];
+
+	if (cached) {
+		return cached[locale] as T;
```
Comment (mwistrand) [inline src/i18n.ts]: Without the `as T`, the TS compiler warns that `Messages` cannot be assigned to type `void | T`; hence the casting.

File: src/i18n.ts [hunk 2]
```diff
@@ -90,6 +90,7 @@ export interface Messages {
 const PATH_SEPARATOR: string = has('host-node') ? global.require('path').sep : '/';
 const VALID_PATH_PATTERN = new RegExp(PATH_SEPARATOR + '[^' + PATH_SEPARATOR + ']+$');
 const contextObjects: LocaleContext<LocaleState>[] = [];
+let bundleMap = {} as BundleMap<Messages>;
```
Comment (kitsonk) [inline src/i18n.ts]: Would it be better to use a `Map` here?

Comment (kitsonk) [inline src/i18n.ts]: Or at least a `Object.create(null)`?

File: src/i18n.ts [hunk 3]
```diff
@@ -305,15 +332,22 @@ function i18n<T extends Messages>(bundle: Bundle<T>, context?: any): Promise<T>
...
 
+	const localePaths = resolveLocalePaths(path, locale, locales);
 	return loadLocaleBundles(localePaths).then((bundles: T[]): T => {
-		return bundles.reduce((previous: T, partial: T, i: number): T => {
+		return bundles.reduce((previous: T, partial: T): T => {
 			const localeMessages = assign({}, previous, partial);
+			let localeCache = bundleMap[bundlePath];
+
+			if (!localeCache) {
+				localeCache = bundleMap[bundlePath] = {};
```
Comment (kitsonk) [inline src/i18n.ts]: Again, maybe `Object.create(null)`?

File: src/i18n.ts (hunk 2 above)
Comment (mwistrand) [inline src/i18n.ts]: A `Map` would be more elegant here.

**Pull request merged**